src/covid19model/visualization/optimization.py

```diff
@@ -0,0 +1,146 @@
+
+    Returns
+    -----------
+
+    Notes
+    -----------
+
+    Example use
+    -----------
+
+
+    """
+
+    # Initialize figure and visualize data
+    # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    # check if ax object is provided by user
+    if ax is None:
+        fig, ax = plt.subplots()
+    # Create shifted index vector using self.extraTime
+    idx = pd.date_range(startDate,freq='D',periods=data[0].size + model.extraTime) - datetime.timedelta(days=model.extraTime)
+    # Plot data
+    for i in range(len(data)):
+        ax.scatter(idx[model.extraTime:],data[i],color="black")
+
+    # Perform simulation(s) and plot model prediction
+    # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    # Compute number of dataseries
+    n = len(data)
+    # Compute simulation time
+    T = data[0].size+model.extraTime-1
+    # Perform simulation(s)
+    if samples is None:
+        k = 0
+        while k < 200:
+            out = model.sim(T,checkpoints=checkpoints)
+            out = out.sum(dim="stratification")
+            for i in range(len(data)):
+                data2plot = out[states[i]].to_array(dim="states").values.ravel()
+                ax.plot(idx,data2plot,linewidth=0.25,alpha=0.2,color="blue")
+            k = k +1
+    else:
+        original_parameters = model.parameters.copy()
+        k=0
+        while k < 100:
+            for key,value in samples:
+                # do random draw and assign to model
+                model.parameters[key] = random.choice(value)
+            # run simulation
+            out = model.sim(T,checkpoints=checkpoints)
+            out = out.sum(dim="stratification")
+            for i in range(len(data)):
+                data2plot = out[states[i]].to_array(dim="states").values.ravel()
+                ax.plot(idx,data2plot)
+            k = k+1
+        # reset parameters
+        model.parameters=original_parameters
+    
+
+    # Attributes
+    if legendText is not None:
+        ax.legend(legendText, loc="upper left", bbox_to_anchor=(1,1))
+    if titleText is not None:
+        ax.set_title(titleText,{'fontsize':18})
+    plt.gca().xaxis.set_major_locator(mdates.DayLocator())
+    plt.gca().xaxis.set_major_formatter(matplotlib.dates.DateFormatter('%d-%m-%Y'))
+    plt.setp(plt.gca().xaxis.get_majorticklabels(),
+        'rotation', 90)
+    ax.set_xlim( idx[model.extraTime-3], pd.to_datetime(idx[-1]+ datetime.timedelta(days=1)))
+    ax.set_ylabel('number of patients')
+    # Hide the right and top spines
+    ax.spines['right'].set_visible(False)
+    ax.spines['top'].set_visible(False)
+    # Only show ticks on the left and bottom spines
+    ax.yaxis.set_ticks_position('left')
+    ax.xaxis.set_ticks_position('bottom')
+    # enable the grid
+    plt.grid(True)
+
+    # limit the number of ticks on the axis
+    ax = _apply_tick_locator(ax)
+
+    if filename is not None:
+        plt.savefig(filename,dpi=600,bbox_inches='tight')
+    if getfig:
+        return fig, ax
```

Comment:
I would drop this section and always return the 'artists' (matplotlib jargon for the points, lines, polygons,... added to a ax) from the function. As such, the pattern becomes:

- user defined fig, ax as she/he wants (with or without subplots,...)
- call function that adds _something_ to an ax object and return the added _something_ (artists)
- user can further customize both her/his ax object and customize the returned `artists`

Hence, reusable building blocks for setting up visualisations. It also follows the good practices defined https://colcarroll.github.io/yourplotlib/